ceph.spec.in

```diff
@@ -224,6 +224,8 @@ BuildRequires:	python%{python3_pkgversion}
 BuildRequires:	python%{python3_pkgversion}-devel
 BuildRequires:	python%{python3_pkgversion}-setuptools
 BuildRequires:	python%{python3_pkgversion}-Cython
+BuildRequires:	python%{python3_pkgversion}-asyncssh
+BuildRequires:	python%{python3_pkgversion}-natsort
```

Comment:
i believe these two python modules are runtime dependencies and are not used when building ceph. per your description 

>  As consequence when a user builds ceph locally and try to use vstart.sh to bootstrap a minimal cluster ceph mgr couldn't load cephadm because of the missing dependencies.

that's exactly your use case. so i'd prefer keep `ceph.spec.in` as it is.

Reply:
@tchaikov well, they are used if the user uses the vstart [https://docs.ceph.com/en/latest/dev/cephadm/developing-cephadm/](vstart) procedure to bootstrap the cluster. While I agree that these are run-time dependencies right now we don't have any "specific" section for the vstart. Please advice an alternative to proceed in this case (If you think that including them will mess the main ceph.spec.in).

Reply:
yeah, we don't have any section for vstart. but ceph.spec is only for building ceph and for running some of its unit tests. it's just *wrong* to dump the *runtime* dependencies as `BuildRequires`.

i don't have a quick answer to your question. but i assume an ordinary user does not use vstart. a developer does though. probably one should improve `install-deps.sh` allowing developer to install more runtime dependencies to address certain needs. but it'd depend on what the developer is working on. and since ceph is a large project, i am afraid there is no simple one-size-fits-all solution. probably another solution is to have a preloaded image...

Reply:
@tchaikov I see, this doesn't seem to have an _easy_ solution. I'll discuss it with the team and see what are the alternatives. We need some easy mechanism for devs who want to use vstart and cephadm.

Reply:
@tchaikov: vstart is essential to Ceph API tests, which runs from build artifacts, not from packages. The interesting thing is that we were already installing `asyncssh` as a build dependency, but not via packages, but implicitly via pip (`mgr/requirements.txt`).

I'd personally prefer to see those dependencies made explicit via `BuildRequires` as well. Those cover 2 different testing scenarios:
* **unit-testing**: pip + tox, quick tests (Ceph built artifacts are mocked),
* **Ceph API testing**: vstart + RPMs. These are the closest ones to a teuthology run, but [providing feedback in 70-min on a per-PR basis](https://jenkins.ceph.com/job/ceph-api/buildTimeTrend) (vs. the 2-step manually-triggered 4-hour shaman+teuthology runs).

BTW: if you check the Dashboard `requirements.txt`, everything there is also specified as a `BuildRequires` dep, so for consistency we should add these two.

Reply:
@tchaikov please, can you have a look to the new updates on this thread. Adding BuildReq seems to be normal for other similar Python modules so it would be good if we can do the same for cephadm. Thanks.

Reply:
> I'd personally prefer to see those dependencies made explicit via `BuildRequires` as well. Those cover 2 different testing scenarios:
> 
>     * **unit-testing**: pip + tox, quick tests (Ceph built artifacts are mocked),
> 
>     * **Ceph API testing**: vstart + RPMs. These are the closest ones to a teuthology run, but [providing feedback in 70-min on a per-PR basis](https://jenkins.ceph.com/job/ceph-api/buildTimeTrend) (vs. the 2-step manually-triggered 4-hour shaman+teuthology runs).

Does `install-deps.sh` get called before both of those?

Reply:
@epuertat i explained the reason why i don't want to approve this change. but none of my concerns is addressed in your comment. but let me try to address your arguments.

> @tchaikov: vstart is essential to Ceph API tests, which runs from build artifacts, not from packages. The interesting thing is that we were already installing `asyncssh` as a build dependency, but not via packages, but implicitly via pip (`mgr/requirements.txt`).
> 

all true. but i am not sure how this is connected to this change.

> I'd personally prefer to see those dependencies made explicit via `BuildRequires` as well. Those cover 2 different testing scenarios:
> 
> * **unit-testing**: pip + tox, quick tests (Ceph built artifacts are mocked),
> * **Ceph API testing**: vstart + RPMs. These are the closest ones to a teuthology run, but [providing feedback in 70-min on a per-PR basis](https://jenkins.ceph.com/job/ceph-api/buildTimeTrend) (vs. the 2-step manually-triggered 4-hour shaman+teuthology runs).

yes. but you are basically repeating Redouane's words. and my concern still holds. i understand the logic of:

```py
if foo.addresses(our_isssue):
  apply(foo)
```

but i don't want to follow it
```py
if foo is wrong:
```

> 
> BTW: if you check the Dashboard `requirements.txt`, everything there is also specified as a `BuildRequires` dep, so for consistency we should add these two.

consistency? yes. correctness? no.

if i had the bandwidth to cleanup `debian/control`, i'd fix them all. but i don't. what i could do is to prevent this from happening again as long as i am the one who is reviewing the change of the same category.

if you insist, we have "<pkg.ceph.check>" profile for debian derivatives and "make_check" condition for rpm packaging for adding dependencies for running "make check". IMHO, it's still wrong, as we don't use this package when running "make check" or "ctest", but it's less wrong.

Reply:
> if you insist, we have "<pkg.ceph.check>" profile for debian derivatives and "make_check" condition for rpm packaging for adding dependencies for running "make check". IMHO, it's still wrong, as we don't use this package when running "make check" or "ctest", but it's less wrong.

@tchaikov that's a good idea, and I have to insist because in fact we do use those packages, but unfortunately we unit test our code with pip versions that don't necessarily match the distro ones (python3-asyncssh 2.9 via pip, 2.7 from CentOS 8, 1.12 in Focal Fossa)... And IMHO that's a bad practice.

@rkachach can you please tag those as @tchaikov suggested? Thanks!

@tchaikov  would you be ok with this then? 🥺 (<--- those meant to be puppy eyes)

Reply:
@epuertat yeah, that'd work =)

adding a `BuildRequires` might mislead our downstream maintainers of Ceph packages. but a build profile on debian and a conditionalized dependency might send them a signal that: this one is not strictly necessary for building Ceph.

Reply:
@tchaikov @epuertat I moved the deps to the `%if 0%{with make_check}` section. Please, take a look and let me know if more changes are needed. Thanks.